### PR TITLE
Add stow task and update pick task

### PR DESCRIPTION
**Pull request by knorth55**
Related to #2110 and #2053 

# How to use

```bash
roslaunch jsk_arc2017_baxter baxter.launch pick:=false
roslaunch jsk_arc2017_baxter setup_for_stow.launch
roslaunch jsk_arc2017_baxter stow.launch
```

# Main Change
- add stow_task files in `jsk_arc2017_baxter`
  - `euslisp/stow-main.l`
  - `euslisp/lib/stow-interface.l`
  - `launch/setup/setup_for_stow.launch`
  - `launch/main/stow.launch`
  - `rvizconfig/stow.rviz`
- add stow_task files in `jsk_arc2017_common`
  - `scripts/stow_json_generator.py`
  - `config/stow_task/shelf_marker.yaml`
  - `data/json/sample_stow_task`
- `node_scripts/state_server.py` support stow_task
- remove slot `bin-boxes-` and replace by `bin-cubes-` in `euslisp/lib/arc-interface.l`
- generalize picking methods for pick task and stow task in `euslisp/lib/arc-interface.l`
- add stow task methods and slots in `euslisp/lib/arc-interface.l`

# Minor Change
- remove method `:ik->bin-entrance` and replace by `ik->bin-center`
- rename method `:recognize-objects-in-bin` -> `:recognize-target-object` in `arc-interface.l`
- rename method `:ik->cardboard-entrance` -> `:ik->cardboard-center`
- reset slot `order` in `pick-interface.l`
- fix typo in `pick_json_generator.py`
- update pick state machine in `pick-main.l`

# State Machine 
## Pick Task
Updated!
arc_pick_state_machine

## Stow Task
arc_stow_state_machine

# Video
https://drive.google.com/open?id=0B5DV6gwLHtyJbkRmbkZxZEhXVjg

### CODE REVIEW

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l [hunk 1]
```diff
@@ -108,24 +114,38 @@
...
+          (sethash arm tote-cubes- (send self :bbox->cube bbox)))
+        (ros::ros-fatal "[:recognize-tote-boxes] arm: ~a Cannot find tote bbox" (arm2str arm)))))
   (:visualize-bins ()
     (let (bins)
-      (dolist (b (send bin-boxes- :list-values))
-        (let ((bin-inside (send self :bbox->cube b))
-              (bin-outside (make-cube (+ (m->mm (send b :dimensions :x)))
-                                      (+ (m->mm (send b :dimensions :y)) 30)
-                                      (+ (m->mm (send b :dimensions :z)) 30)))
+      (dolist (box-inside (send bin-cubes- :list-values))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: Does bin-cubes- have :list-values method?

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l [hunk 2]
```diff
@@ -158,26 +179,44 @@
...
+      (setq graspingp (send* self :pick-object-with-movable-region arm movable-region args))
+      graspingp))
+  (:pick-object-with-movable-region
+    (arm movable-region &key (n-trial 1) (n-trial-same-pos 1) (do-stop-grasp nil))
+    (let (graspingp avs object-index)
+      ;; TODO: object-index is set randomly
+      (setq object-index (random (length (gethash arm object-boxes-))))
+      (setq obj-pos
+            (send self :get-object-position arm movable-region :object-index object-index))
+      (ros::ros-info "[:pick-object-with-obj-pos] arm:~a approach to the object" arm)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: pick-object-with-movable-region?

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l [hunk 3]
```diff
@@ -279,50 +298,31 @@
...
+    (let (bin-cube bin-coords)
+      (setq bin-cube (gethash bin bin-cubes-))
+      (setq bin-coords (send bin-cube :copy-worldcoords))
       (send bin-coords :rotate (aref rpy 0) :z)
       (send bin-coords :rotate (aref rpy 1) :y)
       (send bin-coords :rotate (aref rpy 2) :x)
+      (send bin-coords :translate
+            (v+ (float-vector
+                  0.0 0.0 (/ (z-of-cube bin-cube) 2.0))
+                 offset) :world)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: "z" of `z-of-cube` is repesented in local coodinates, isn't it?
Then please translate in local.
Offset can be applied in whether local or world.

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l [hunk 4]
```diff
@@ -333,6 +333,21 @@
...
+    (let (tote-cube tote-coords)
+      (setq tote-cube (gethash arm tote-cubes-))
+      (setq tote-coords (send tote-cube :copy-worldcoords))
+      (send tote-coords :rotate (aref rpy 0) :z)
+      (send tote-coords :rotate (aref rpy 1) :y)
+      (send tote-coords :rotate (aref rpy 2) :x)
+      (send tote-coords :translate
+            (v+ (float-vector
+                  0.0 0.0 (/ (z-of-cube tote-cube) 2.0))
+                 offset) :world)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: "z" of `z-of-cube` is repesented in local coodinates, isn't it?
Then please translate in local.
Offset can be applied in whether local or world.

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l [hunk 5]
```diff
@@ -374,22 +404,19 @@
...
       (format nil "/bin_contents/~A" (string-upcase (symbol2str bin)))))
   (:add-bin-scene (bin)
-    (let ((cube (send self :bbox->cube (gethash bin bin-boxes-)))
+    (let ((cube (gethash bin bin-cubes-))
           (base-name (send (send *baxter* :base_lk) :name)))
       (send *co* :add-object cube :frame-id base-name
             :relative-pose (send cube :copy-worldcoords)
-            :object-id (format nil "bin_~A" (string-upcase (symbol2str bin))))
-      (sethash bin bin-cubes- cube)))
+            :object-id (format nil "bin_~A" (string-upcase (symbol2str bin))))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: (string-upcase (symbol2str ~) -> (symbol-string ~)

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    (send *ri* :wait-interpolation))
+  (:place-object (arm)
+    (let (dropped)
+      (ros::ros-info "[main] ~a, place object in bin ~a" arm target-bin)
+      (send self :add-tote-scene arm)
+      (send *baxter* :head_pan :joint-angle (if (eq arm :larm) 80 -80))
+      (send *ri* :angle-vector
+            (send self :ik->bin-center arm target-bin
+                  :offset #f(0 0 300) :rotation-axis :z :use-gripper t)
+            4000 (send *ri* :get-arm-controller arm) 0)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: if you can, please use `scale`

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+      (setq dropped (not (send *ri* :graspingp arm)))
+      (if dropped
+        (progn
+          (ros::ros-error "[main] arm ~a: dropped object" arm)
+          (send *ri* :stop-grasp arm))
+        (progn
+          (ros::ros-info-green "[main] arm ~a: place object ~a in bin ~a" arm target-obj target-bin)
+          (send self :delete-bin-scene target-bin)
+          (send *ri* :angle-vector-raw (send *baxter* arm :move-end-pos #f(0 0 -300) :world)
+                2000 (send *ri* :get-arm-controller arm) 0)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: if you can, please use `scale`

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+          (send *ri* :wait-interpolation)
+          (send *ri* :stop-grasp arm) ;; release object
+          (send self :spin-off-by-wrist arm :times 10)
+          (send *ri* :wait-interpolation)
+          (send *ri* :angle-vector-raw (send *baxter* arm :move-end-pos #f(0 0 300) :world)
+                1000 (send *ri* :get-arm-controller arm) 0)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: if you can, please use `scale`

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l (hunk 1 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: Yes, it is hash-table

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l (hunk 6 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: currently, `scale` is not supported by `moveit` method.
if you set tm as `:fast`, `angle-vector-raw` will be executed instead.

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l (hunk 7 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: https://github.com/start-jsk/jsk_apc/pull/2138#discussion_r121271758

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l (hunk 8 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: https://github.com/start-jsk/jsk_apc/pull/2138#discussion_r121271758

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l (hunk 1 above)
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: Understood. Sorry

File: jsk_arc2017_baxter/euslisp/lib/stow-interface.l (hunk 6 above)
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/stow-interface.l]: Opps, I forgot that. Sorry

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l (hunk 5 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: Ok

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l (hunk 4 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: you are right

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l (hunk 3 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: you are right

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l (hunk 2 above)
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: typo

**Comment by knorth55** [comment]
Updated

File: jsk_arc2017_baxter/launch/main/stow.launch [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    <node unless="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="left_hand_main"
+          args="&quot;(progn (stow-init :ctype :larm-controller :moveit nil) (stow-mainloop :larm))&quot;"
+          respawn="true" output="screen">
+      <env name="DISPLAY" value="" />
+    </node>
+
+    <node unless="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="right_hand_main"
+      args="&quot;(progn (stow-init :ctype :rarm-controller :moveit nil) (stow-mainloop :rarm))&quot;"
+      respawn="true" output="screen">
+      <env name="DISPLAY" value="" />
```
Comment (pazeshun) [inline jsk_arc2017_baxter/launch/main/stow.launch]: indent

File: jsk_arc2017_baxter/launch/main/stow.launch [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    <node if="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="left_hand_main"
+          args="&quot;(progn (stow-init :ctype :larm-controller :moveit t) (stow-mainloop :larm))&quot;"
+          respawn="true" output="screen">
+      <env name="DISPLAY" value="" />
+    </node>
+
+    <node if="$(arg moveit)" pkg="jsk_arc2017_baxter" type="stow-main.l" name="right_hand_main"
+      args="&quot;(progn (stow-init :ctype :rarm-controller :moveit t) (stow-mainloop :rarm))&quot;"
+      respawn="true" output="screen">
+      <env name="DISPLAY" value="" />
```
Comment (pazeshun) [inline jsk_arc2017_baxter/launch/main/stow.launch]: indent

**Comment by wkentaro** [comment]
@knorth55 @pazeshun We need to decide how to classify (recognize) grasped object (or recognized before grasping).
My suggestion is below pipeline for version 2 (this PR is version 1) of stow task system.

1. Use FCN to grasp graspable objects.
2. Use FCN to recognize grasped object.
3. Stow to shelf.

So, what we need is an external camera or use of inverse hand camera to see current grasping hand.
For this, I suggest use of inverse hand camera and do in-hand recognition, because it does not need calibration of external camera, and it is a familiar way for us.
Could one of you tackle this?

**Comment by pazeshun** [comment]
Ok, but both of us have tasks of high priority(me: #2111 , #2123 @knorth55: #2017)
Maybe the one who has more time than the other will tackle (Currently I can't decide I'll do)

**Comment by knorth55** [comment]
In-hand recognition, I want to use external camera in the last version, but for version 2, it is ok.
If I reuse my bthesis code, I can do it quickly.

**Comment by knorth55** [comment]
Anyway, this PR is rebased.

**Comment by knorth55** [comment]
plz merge this

**Pull request merged**